SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ComputeEffects2.swift

```diff
@@ -0,0 +1,271 @@
+//===--- ComputeEffects.swift - Compute function effects ------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+fileprivate typealias Selection = ArgumentEffect.Selection
+fileprivate typealias Path = ArgumentEffect.Path
+
+/// Computes effects for function arguments.
+///
+/// For example, if an argument does not escape, adds a non-escaping effect,
+/// e.g. "[escapes !%0.**]":
+///
+///   sil [escapes !%0.**] @foo : $@convention(thin) (@guaranteed X) -> () {
+///   bb0(%0 : $X):
+///     %1 = tuple ()
+///     return %1 : $()
+///   }
+///
+/// The pass does not try to change or re-compute _defined_ effects.
+/// Currently, only escaping effects are handled.
+/// In future, this pass may also add other effects, like memory side effects.
+let computeEffects2 = FunctionPass(name: "compute-effects2", {
+  (function: Function, context: PassContext) in
+  
+  struct Visitor : EscapeInfoWalkerVisitor {
```

Comment:
The visitor structs make the code a little bit unreadable.
To improve the readability, I suggest to create an `extension EscapeInfoWalker` containing a separate member function for each visitor use case. And put the visitor structs into those functions. E.g. for this use case:
```
extension EscapeInfoWalker {
  func canArgumentEscape(_ arg: FunctionArgument) -> Bool {
    struct Visitor : EscapeInfoWalkerVisitor {
      ...
    }
    return escapeInfo.isEscapingWhenWalkingDown(object: arg, path: Path(.anything), visitor: Visitor())
  }
}
```